lading_payload/src/block.rs

```diff
@@ -463,65 +485,73 @@ fn chunk_bytes<const N: usize>(
 /// Function will panic if the `serializer` signals an error. In the future we
 /// would like to propagate this error to the caller.
 #[inline]
-#[tracing::instrument(skip(rng, serializer, block_chunks))]
+#[tracing::instrument(skip_all)]
 fn construct_block_cache_inner<R, S>(
     mut rng: &mut R,
     serializer: &S,
     block_chunks: &[u32],
+    total_bytes: u32,
 ) -> Result<Vec<Block>, SpinError>
 where
     S: crate::Serialize,
     R: Rng + ?Sized,
 {
-    let mut block_cache: Vec<Block> = Vec::with_capacity(block_chunks.len());
-    let mut block_iter = block_chunks.iter().peekable();
-    let mut construct_block_failures = 0;
-    let max_construct_block_failures: usize = block_chunks.len() * 5;
-    while let Some(peeked_block) = block_iter.peek() {
-        assert!(construct_block_failures <= max_construct_block_failures, "Block construction failed {construct_block_failures} times consecutively while trying to construct block of size {peeked_block}. Failing the program. Check payload generation parameters.");
-        // Consume the block only if a block is successfully constructed
-        // otherwise retry the block size until it succeeds.
-        block_iter.next_if(|block_size| {
-            if let Ok(block) = construct_block(&mut rng, serializer, **block_size) {
-                construct_block_failures = 0; // reset failure counter on success
+    let mut block_cache: Vec<Block> = Vec::with_capacity(128);
+    let mut bytes_remaining = total_bytes;
+    let mut min_viable_size = 0;
+    let max_viable_size: u32 = *block_chunks
+        .iter()
+        .max()
+        .expect("empty block chunks, catastrophic bug");
+
+    // Build out the blocks.
+    //
+    // Our strategy here is to keep track of the minimal viable size of a block
+    // -- `min_viable_size` -- as the "floor" for block sizes. Because the
+    // serialization format varies we can't know what the floor actually is
+    // until runtime. We take the user-provided blocks and the total byte
+    // objective and iterate over these, dynamically constructing new block
+    // sizes +/- 20% if the block construction fails because of insufficient
+    // space to serialize.
+    while bytes_remaining > 0 {
+        let block_size = rng.gen_range(min_viable_size..=max_viable_size);
+
+        match construct_block(&mut rng, serializer, block_size) {
+            Ok(block) => {
+                bytes_remaining = bytes_remaining.saturating_sub(block.total_bytes.get());
                 block_cache.push(block);
-                true
-            } else {
-                construct_block_failures += 1;
-                false
             }
-        });
+            Err(SpinError::EmptyBlock) => {
+                min_viable_size = block_size;
```

Comment:
Due to the random nature of the generators, the same request of N bytes may fail once and then succeed the next time. I don't know how to work around that fact in the algorithm, but I think we should at least have a check here to only grow the `min_viable_size`.

```suggestion
                min_viable_size = std::min(block_size, min_viable_size);
```

Reply:
And/or adjust randomly per the +-20% idea

Reply:
Ah yeah that is an interesting point. Because the `block_size` is  [min_viable_size, max_viable_size] we're guaranteed that it's always >= `min_viable_size`, so I think a `min` operation here is not called for. The minimum viable size will never fall below its current level. There will be some iterations where it does not change but I think that's acceptable.

Reply:
Reintroduced random fluctuation in [ecf5db3](https://github.com/DataDog/lading/pull/886/commits/ecf5db3d3d8e0c600b6208ea900d285aa16a79ae). We achieve a tighter lower bound on the minimum block size -- 62B in one run where 1.1Kb had been the norm.